autoload/maktaba/keymappingspec.vim

```diff
@@ -0,0 +1,130 @@
+  let l:mode = get(a:, 1, '')
+
+  let l:spec = {
+      \ '_lhs': a:lhs,
+      \ '_rhs': a:rhs,
+      \ '_mode': l:mode,
+      \ '_args': ['<unique>'],
+      \ '_is_noremap': 1,
+      \ '_is_bufmap': 0,
+      \ 'WithArgs': function('maktaba#keymappingspec#WithArgs'),
+      \ 'WithRemap': function('maktaba#keymappingspec#WithRemap'),
+      \ 'Map': function('maktaba#keymappingspec#MapSelf'),
+      \ 'MapOnce': function('maktaba#keymappingspec#MapSelfOnce'),
+      \ 'MapOnceWithTimeout': 
+          \ function('maktaba#keymappingspec#MapSelfOnceWithTimeout'),
+      \ }
+  return l:spec
+endfunction
+
+
+""
+" @dict KeyMappingSpec
+" Add {args} as |map-arguments| to spec (appended to any already configured on
+" spec), or overwrite instead of appending if [overwrite] is passed and is true.
+" {args} is a list of strings in the literal syntax accepted by |:map|, such as
+" `['<buffer>', '<nowait>']`.
+" @default overwrite=0
+function! maktaba#keymappingspec#WithArgs(args, ...) dict abort
+  let l:should_overwrite = maktaba#ensure#IsBool(get(a:, 1, 0))
+  let l:spec = copy(self)
+  if l:should_overwrite
+    " TODO: Test this case
+    let l:spec._args = a:args
+  else
+    let l:spec._args += a:args
+  endif
+  if index(a:args, '<buffer>') >= 0
+    let l:spec._is_bufmap = 1
+  endif
+  return l:spec
+endfunction
+
+
+""
+" @dict KeyMappingSpec
+" Configure whether spec should have nested/recursive mappings {enabled}.
+" @throws WrongType
+function! maktaba#keymappingspec#WithRemap(enabled) dict abort
+  let l:spec = copy(self)
+  let l:spec._is_noremap = !maktaba#ensure#IsBool(a:enabled)
+  return l:spec
+endfunction
+
+
+""
+" @dict KeyMappingSpec.Map
+function! maktaba#keymappingspec#MapSelf() dict abort
+  let l:keymap = maktaba#keymapping#PopulateFromSpec(self)
+  call l:keymap._DoMap()
+  return l:keymap
+endfunction
+
+
+""
+" @dict KeyMappingSpec.MapOnce
+" Define a buffer-local one-shot vim mapping from spec that will only trigger
+" once and then unmap itself.
+"
+" Not supported for recursive mappings.
+" @throws NotImplemented if used with `WithRemap(1)`
+function! maktaba#keymappingspec#MapSelfOnce() dict abort
+  let l:keymap = maktaba#keymapping#PopulateFromSpec(self)
+  call l:keymap._DoMapOnce()
+  return l:keymap
+endfunction
+
+
+""
+" @dict KeyMappingSpec.MapOnceWithTimeout
+" Define a short-lived vim mapping from spec that will only trigger once and
+" will also expire if 'timeoutlen' duration expires with 'timeout' setting
+" active.
+"
+" This is useful to detect if the user presses a key immediately after something
+" else happens, and respond with a particular behavior.
+"
+" It can also be used for an improved version of vim's |map-ambiguous| behavior
+" when one mapping is a prefix of another. You can create a prefix mapping that
+" does one thing immediately and then a different follow-up behavior on another
+" keystroke.
+"
+" For example, this defines a mapping that immediately unfolds one level but
```

Comment:
I'm really surprised that this works, but we have tests, so sgtm!

Reply:
Yeah, actually it looks like I had a typo in the args there I just fixed.